src/Snap/Internal/Http/Server/Parser.hs

```diff
@@ -173,18 +177,35 @@ parseRequest input = do
                     in (Just $! host, uri)
               | otherwise = (Nothing, s)
 
-    pVer s = if "HTTP/" `S.isPrefixOf` s
-               then pVers (S.unsafeDrop 5 s)
-               else (1, 0)
+    -- RFC 7230 section 2.6 defines
+    --
+    --   HTTP-version = HTTP-name "/" DIGIT "." DIGIT
+    --   HTTP-name    = %x48.54.54.50 ; "HTTP", case-sensitive
+    --
+    pVer s = case bsStripHttpPrefix s of
+      Nothing -> return (1, 0)
+      Just "1.1" -> return (1, 1)
+      Just "1.0" -> return (1, 0)
+      Just vstr
+        | [mjs,'.',mns] <- S.unpack vstr
+        , Just mj <- digitToInt mjs
```

Comment:
Same here with calling `digitToInt`. We are more okay with a misparse returning version '0.0' than we are with adding another *x* bytes of allocation overhead to this function

Reply:
even if this is the unhappy path, which would only occur for "bad" clients?

Reply:
e.g. `bsStripHttpPrefix` that returns a `Maybe` - that's in the hot path, so we'd prefer to return `ByteString` there to save three words and a pointer indirection. Null string is as good as `Nothing` there. `case`ing on the resulting bytestrings using string equality in the hot path is probably faster than splitting on `.` like I was doing before, though.